### PR TITLE
[SYCLomatic] Enable migration of CUBLASLT: EPILOGUE_BIAS, EPILOGUE_GELU, EPILOGUE_GELU_AUX, EPILOGUE_GELU_AUX_BIAS

### PR DESCRIPTION
**Pull request by abhilash1910**
@zhiweij1 @zhimingwang36

### CODE REVIEW

File: clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp [hunk 1]
```diff
@@ -1024,7 +1026,14 @@ inline sycl::event matmul(descriptor_ptr handle, matmul_desc_ptr compute_desc,
 
   if (compute_desc->_epilogue != epilogue_t::nop) {
     ::dnnl::post_ops matmul_ops;
-    matmul_ops.append_eltwise(::dnnl::algorithm::eltwise_relu, 0.f, 0.f);
+    if (compute_desc->_epilogue == epilogue_t::relu) {
+      matmul_ops.append_eltwise(::dnnl::algorithm::eltwise_relu, 0.f, 0.f);
+    } else if (compute_desc->_epilogue == epilogue_t::bias) {
+      matmul_ops.append_binary(::dnnl::algorithm::binary_add, bias_md);
+    } else if (compute_desc-> epilogue == epilogue::gelu_aux_bias) {
+      matmul_ops.append_eltwise(::dnnl::algorithm::eltwise_gelu_erf, 0.f, 0.f);
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp]: 1. `CUBLASLT_EPILOGUE_GELU_AUX_BIAS` needs to output the `bias` result to `CUBLASLT_MATMUL_DESC_EPILOGUE_AUX_POINTER`. But here, your implementation does not.
2. Please write an E2E test to validate your implementation. I tried something similar before but encountered a runtime error or an incorrect result.

**Review by zhiweij1 — CHANGES_REQUESTED** [review]
Please refine

File: clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp (hunk 1 above)
Comment (abhilash1910) [reply clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp]: Ok. Please help to review.

File: clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp [hunk 2]
```diff
@@ -1024,7 +1026,15 @@ inline sycl::event matmul(descriptor_ptr handle, matmul_desc_ptr compute_desc,
...
+      matmul_ops.append_binary(::dnnl::algorithm::binary_add, bias_md);
+      compute_desc->set_attribute(matmul_desc_t::attribute::epilogue_aux_pointer, bias_mem);
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp]: We need to copy the data in `bias_mem` to `_epilogue_aux_pointer`. Meanwhile, please also consider the data layout (row_major/col_major) and the leading dimension (`CUBLASLT_MATMUL_DESC_EPILOGUE_AUX_LD`).

And according to the cublaslt doc, `CUBLASLT_MATMUL_DESC_BIAS_POINTER` is also related to this epilogue. Please check the usage.

**Review by zhiweij1 — CHANGES_REQUESTED** [review]
Please refine

**Comment by zhiweij1** [comment]
Also could you please prepare an E2E test first? like https://github.com/oneapi-src/SYCLomatic-test/blob/SYCLomatic/features/feature_case/cublasLt/matmul.cu .
Make sure your test case (cuda code) can run pass on NV HW. Then migrate it to sycl. Compile the sycl code within your PR, make sure the result is same as the cuda code output.

File: clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp (hunk 2 above)
Comment (abhilash1910) [reply clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp]: I directly added matrix memcpy  to copy epilogue_aux_pointer contents from bias_mem pointer, using the epilogue_aux_ld is and new_ldc (bias ld). This usage is for matrix memcpy, and will try to see if tests are ok .
Please suggest in meantime.

File: clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp [hunk 3]
```diff
@@ -1027,7 +1029,17 @@ inline sycl::event matmul(descriptor_ptr handle, matmul_desc_ptr compute_desc,
...
     ::dnnl::post_ops matmul_ops;
-    matmul_ops.append_eltwise(::dnnl::algorithm::eltwise_relu, 0.f, 0.f);
+     if (compute_desc->_epilogue == epilogue_t::relu) {
+      matmul_ops.append_eltwise(::dnnl::algorithm::eltwise_relu, 0.f, 0.f);
+    } else if (compute_desc->_epilogue == epilogue_t::bias) {
+      matmul_ops.append_binary(::dnnl::algorithm::binary_add, bias_md);
+    } else if (compute_desc->_epilogue == epilogue_t::gelu_aux_bias) {
+      matmul_ops.append_eltwise(::dnnl::algorithm::eltwise_gelu_erf, 0.f, 0.f);
+      matmul_ops.append_binary(::dnnl::algorithm::binary_add, bias_md);
+      dpct::blas::matrix_mem_copy(matmul_desc_t::attribute::epilogue_aux_pointer, bias_mem,
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp]: matmul_desc_t::attribute::epilogue_aux_pointer is a enum, not a pointer. bias_mem is a pointer of dnnl::memory, not a pointer to data.

File: clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp [hunk 4]
```diff
@@ -1024,10 +1029,27 @@ inline sycl::event matmul(descriptor_ptr handle, matmul_desc_ptr compute_desc,
...
   if (compute_desc->_epilogue != epilogue_t::nop) {
     ::dnnl::post_ops matmul_ops;
-    matmul_ops.append_eltwise(::dnnl::algorithm::eltwise_relu, 0.f, 0.f);
+     if (compute_desc->_epilogue == epilogue_t::relu) {
+      matmul_ops.append_eltwise(::dnnl::algorithm::eltwise_relu, 0.f, 0.f);
+    }  else if (compute_desc->_epilogue == epilogue_t::gelu) {
+      matmul_ops.append_eltwise(::dnnl::algorithm::eltwise_gelu_erf, 0.f, 0.f);
+    }  else if (compute_desc->_epilogue == epilogue_t::gelu_aux) {
+      matmul_ops.append_eltwise(::dnnl::algorithm::eltwise_gelu_erf, 0.f, 0.f);
+      dpct::blas::matrix_mem_copy(compute_desc->_epilogue_aux_pointer, new_c,
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp:1047]: I forget one issue. This `dpct::blas::matrix_mem_copy` does not return sycl::event. So we cannot control the dependency chain when using non-USM.

File: clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp [hunk 5]
```diff
@@ -1024,10 +1029,27 @@ inline sycl::event matmul(descriptor_ptr handle, matmul_desc_ptr compute_desc,
...
   }
-
+  sycl::queue &queue = ::dpct::cs::get_default_queue();
   if (compute_desc->_epilogue != epilogue_t::nop) {
     ::dnnl::post_ops matmul_ops;
-    matmul_ops.append_eltwise(::dnnl::algorithm::eltwise_relu, 0.f, 0.f);
+     if (compute_desc->_epilogue == epilogue_t::relu) {
+      matmul_ops.append_eltwise(::dnnl::algorithm::eltwise_relu, 0.f, 0.f);
+    }  else if (compute_desc->_epilogue == epilogue_t::gelu) {
+      matmul_ops.append_eltwise(::dnnl::algorithm::eltwise_gelu_erf, 0.f, 0.f);
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp]:
```suggestion
      matmul_ops.append_eltwise(::dnnl::algorithm::eltwise_gelu_tanh, 0.f, 0.f);
```

File: clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp (hunk 4 above)
Comment (zhiweij1) [reply clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp:1047]: Another issue is new_c is the bias, not the result of (A*B+C). I did not find there is a method to get that immediate result when using oneDNN post-op.
Maybe we need to use a separate oneDNN elewise primitive.

File: clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp [hunk 6]
```diff
@@ -1024,10 +1029,27 @@ inline sycl::event matmul(descriptor_ptr handle, matmul_desc_ptr compute_desc,
...
+    }  else if (compute_desc->_epilogue == epilogue_t::gelu_aux) {
+      matmul_ops.append_eltwise(::dnnl::algorithm::eltwise_gelu_erf, 0.f, 0.f);
+      dpct::blas::matrix_mem_copy(compute_desc->_epilogue_aux_pointer, new_c,
+                                  compute_desc->_epilogue_aux_ld, new_ldc, m, n,
+                                  sizeof(size_t) , dpct::device_to_device, queue);
+    } else if (compute_desc->_epilogue == epilogue_t::bias) {
+      matmul_ops.append_binary(::dnnl::algorithm::binary_add, bias_md);
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp:1052]: bias_md represents the C in the D=A*B+C, not the extra bias in epilogue.

File: clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp (hunk 4 above)
Comment (abhilash1910) [reply clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp:1047]: Yes correct, for the second point, I see it is raised in JIRA. Currently I could not find a way to extract immediate result of A*B+C for the aux memcpy. 

For the first case, do I have to force use USM always then ? What do you suggest ?

File: clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp (hunk 6 above)
Comment (abhilash1910) [reply clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp:1052]: Need to create new bias descriptor. A*B+C+bias.?

File: clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp (hunk 4 above)
Comment (zhiweij1) [reply clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp:1047]: I think we can do something like:
```
    sycl::event copy_e;
    size_t size_of_element =
        dpct::detail::library_data_size[static_cast<unsigned int>(
            d_desc->_type)] /
        8;
    if (compute_desc->_epilogue_aux_ld == d_desc->_ld) {
      size_t copy_size =
          size_of_element * ((n - 1) * (size_t)(d_desc->_ld) + m);
      copy_e = ::dpct::cs::memcpy(
          *q_ptr, (void *)(compute_desc->_epilogue_aux_pointer), (void *)new_d,
          copy_size, ::dpct::cs::memcpy_direction::automatic, {matmul_prim_event});
    } else {
      copy_e = ::dpct::cs::merge_events(
          *q_ptr,
          ::dpct::cs::memcpy(*q_ptr, compute_desc->_epilogue_aux_pointer, new_d,
                             size_of_element * compute_desc->_epilogue_aux_ld,
                             size_of_element * d_desc->_ld, size_of_element * m,
                             n, ::dpct::cs::memcpy_direction::automatic, {matmul_prim_event}));
    }
```
While currently the `::dpct::cs::memcpy` cannot accept dependency events and there is no `::dpct::cs::merge_events`.
We need also adjust/add those 2 APIs.

Comment (abhilash1910) [reply clang/runtime/dpct-rt/include/dpct/blas_gemm_utils.hpp:1047]: Yes adjusted changes after discussion, please help to review. Thanks

**Comment by abhilash1910** [comment]
Transfer to @zhiweij1 . Thanks

**Pull request closed without merging**